### PR TITLE
Color Describer

### PR DESCRIPTION
**Pull request by Kurfursten**
Instead of being able to identify exactly what is in a container, when you look at it you will instead see a description of the color mixed within. This means that you can hide reagents.
* Wearing science goggles will let you examine reagents like you used to
* Smart containers (like the bartender shaker) still allow you to examine reagents
* (TODO) Silicons can see reagents with an upgrade
* (TODO) Ghosts can see reagents

untitled

Buggy behavior:
* Dispensers (like fuel tanks) seem to override examine or something because they are not affected as of yet.

🆑 
* rscadd: Spacemen can no longer identify exactly which reagents are in a container unless they have scientific goggles.

### CODE REVIEW

**Comment by ShiftyRail** [comment]
I would describe medicine colour as "pinkish" rather than grey tbh.

**Comment by aceedex** [comment]
Add the functionality to the medhud as well tbh

**Comment by Kurfursten** [comment]
Under my system, a color is considered gray if it is no more than twice the greatest color as least color. So for bicaridine: 200, 165, 220; 220/2=110<165. It's grayish because all three colors are mostly similar.

The good thing though is that you are free to add more customization later on if you want to tweak it yourself. "Pinkish" is not a supported color. Right now this covers:
* White, gray, black
* Red, green, blue
* Yellow, purple, cyan
* Brown, orange

Brown and orange got "snowflaked" in because we have so many brown and orange things, but you could definitely add more precision in other ways.

File: code/modules/mob/living/carbon/monkey/inventory.dm [hunk 1]
```diff
@@ -125,3 +125,6 @@
 
 /mob/living/carbon/monkey/reversestrip_time()
 	return MONKEY_REVERSESTRIP_DELAY
+
+/mob/living/carbon/monkey/can_see_reagents()
+	return glasses && istype(glasses,/obj/item/clothing/glasses/science)
```
Comment (DamianX) [inline code/modules/mob/living/carbon/monkey/inventory.dm:130]: use `is_wearing_item`

File: code/modules/mob/living/silicon/silicon.dm [hunk 2]
```diff
@@ -28,6 +28,9 @@
 /mob/living/silicon/GetAccess()
 	return get_all_accesses()
 
+/mob/living/silicon/can_see_reagents()
+	return TRUE //Is this creep?
```
Comment (DamianX) [inline code/modules/mob/living/silicon/silicon.dm:32]: yes

File: code/modules/mob/living/carbon/human/inventory.dm [hunk 3]
```diff
@@ -497,3 +497,6 @@
 	if(active_only && istype(get_inactive_hand(),/obj/item/device/multitool))
 		return get_inactive_hand()
 	return null
+
+/mob/living/carbon/human/can_see_reagents()
+	return glasses && istype(glasses,/obj/item/clothing/glasses/science)
```
Comment (DamianX) [inline code/modules/mob/living/carbon/human/inventory.dm:502]: use is_wearing_item

**Comment by SonixApache** [comment]
next up: flavors

**Comment by DrSnips** [comment]
While the changes are somewhat reasonable I have to ask why should this be a thing? So you can hide reagents? Who does this benefit? Traitors? I dont recall the last time I really wanted to hide a reagent containers contents from someone besides as bartender. Food doesnt let you see reagents already so that hides them if you want to trick someone into taking chems.

**Comment by strawberrytan** [comment]
I never liked the ability to mass spectrometise bar drinks down to the unit with the naked eye

**Comment by Kurfursten** [comment]
You're not looking at the bigger picture here, Snips. We've seen floor food and floor pills, sure. This allows for *floor cocktails*. Also just better immersions in general. And because we've been sued by Dan Corp for revealing the secret recipe to Mann's Drink every time someone shift clicks it.

In general this benefits both antag and nonantag bartenders, because no more
>It Contains
>25 units of beer
>4.888888888888888 units of syndicate bomb
>1.111111111111111 units of space cola

Instead
>It's about 60% full of a yellow reagent

I suppose you could also say it also helps further define the role of chemist vs medical doctor (though maybe not since there's a lot of support for adding viewing to the medhud, which I won't fight), it takes away the "meta" of carrying a mortar to check the contents of food with on-the-go-grinding, and also has potential for other mischief like replacing a welder tank with space cola and confusing people since they're both red.

**Comment by DrSnips** [comment]
I'd rather not have chem become "It's a fucking mystery". Say something like sugar and flower for chefs. IRL you can tell the difference between the two but now IC you aren't able to tell the difference.
If you're going to add it make it so they are seperated by something other than just colors like solids liquids and gas.

**Comment by I-VAPE-VOX-CLOACA-EVERY-DAY-OF-MY-LIFE** [comment]
I guess there could easily be "identifiable" reagents like flour and sugar that replace the color-esque description with "it seems to be 75% full of flour"

**Comment by Kurfursten** [comment]
We can do states although
* I don't know when it would even matter since 90% of the time you're either working with 'flour and sugar' which are both solids or 'bicaridine and dexalin' / 'beer and poison wine' which are all liquids - so I'm not sure it will in practice actually give you much more information
* It will draw more attention to the fact that we have several reagents with a state of GAS which makes no sense in an open beaker (or soda can) like oxygen

That being said it would be easy to instead of saying "white reagents" say "white powder" (flour, sugar, etc) / "white suspension" (solid and liquid present, eg flour and milk) / "white liquid" (milk, water, etc). What even to call liquid and gas together...? "white effervescence"?

Fortunately for chefs this doesn't change anything about the display of a microwave or their condimaster so they can still check reagents before hitting go or if they accidentally mix their bags

**Comment by DrSnips** [comment]
While I could understand the removal of the chem amount and have them based on a percent unless you have a mcguffin I just don't like the idea behind them being super obscure now. This makes everyone's life more difficult than it needs to be and has no positive impact on the game. Even with traitors slipping stuff in bar drinks Id rather have it so pills don't get crushed into drink but rather they sit in the drink and upon drinking it you also eat the pill if one was in it. This is the only practical application I can think of besides MUH IMMERSIONS.

**Comment by Arthurdentist** [comment]
I really don't like this change. Its silly that spessmen can immediately know how much of something is in a beaker down to the precise units but at the same time, it would be even sillier if spessmen couldn't tell the difference between blood and ketchup.

**Comment by Exxion** [comment]
I'd been thinking about a system like this for a couple years. Personally I think it should have more information than just color, like "brown, cloudy" or "white, opaque" or "white powder"
But that would require adding that info to each reagent

**Comment by Arthurdentist** [comment]
I feel like if the goal is to make it so that you can hide reagents there is probably a more sensible way of doing it. What if reagents didn't show up if they were in 'trace' amounts? I feel like a spessman should be able to tell if his drink is 99.9% poison, but when mixing poison into a drink, the point is to mask the look and taste so they don't notice.

**Comment by SonixApache** [comment]
should definitely replace 'reagent' with liquid, solution, powder, etc.

**Comment by Chazach** [comment]
why do silicons get to know

**Comment by Riekland** [comment]
>gray

>a

**Comment by Kurfursten** [comment]
Silicons can't wear glasses and it's less weird that a robot can tell the chemical composition of a fluid by sight.

I thought about restricting it to medborgs though.

**Comment by DamianX** [comment]
You could make it a separate board that can fit multiple (all?) modules. Think about the service borgs, too.

**Comment by Kurfursten** [comment]
Doesn't seem like a bad idea

**Comment by jknpj** [comment]
Make it a quirk_flag

**Comment by DrSnips** [comment]
This PR keeps getting worse and worse.

**Comment by I-VAPE-VOX-CLOACA-EVERY-DAY-OF-MY-LIFE** [comment]
Thinking about it some more, I think telling you "it seems to be mostly [reagent]" seems to accomplish the same thing as "it seems to be mostly [color]" without making you play color Guess Who

**Comment by Kurfursten** [comment]
The nice thing about colors is that it's very organic without obvious breakpoints. The amount it shifts the color varies based on what you're pouring in and how much. So it's a lot easier to mix welding fuel into a soda, for example, because they're both red. I think that's mechanically very interesting. And remember the chef does have a reagent scanner on his PDA that is criminally underused. Later tonight I'll sprite a bartender's visor and hilarious chef goggles that might be things service could ask science to print them. Reminds me of that HydroHUD I've wanted to make for years #11496



I do think the idea of including states of matter is good though - the only ones we lack a name for is "gas + solid" and "all three states present". Any ideas?

**Comment by Durgen** [comment]
I like restricting magic chemical vision to sci goggles. It gives them a novel use. Adding the functionality to med goggles would just make sci goggles a shit version of med goggles.

**Comment by DrSnips** [comment]
I still don't like this but to help improve things a bit you could format it like this.
>The beaker contains
A (color) (state of color)
and a (color) (state of color)

So if you have something like hydrogen and salt you would have
>The beaker contains:
A (color of hydrogen here) gas
and a white solid.

**Comment by DrSnips** [comment]
While goggles are nice and all not everyone will have the ability to wear them at all times. You will keep needing to make exceptions like the chef or all of medical so in the end who is this change for again? So you can sneak a bit of welderfuel into soda cans? Id rather have the pills not being broken down when put in beakers so you can "hide" them in it if the only thing you really wanted out of these changes was to hide stuff in drink for griff reasons.

**Comment by Kurfursten** [comment]
That doesn't make sense in most contexts though, where it's "mixing". That's why when you pour two liquids together it becomes a blend of their colors. You aren't seeing a red liquid and a green liquid, you're just seeing a yellow liquid. Similarly it makes things like "A red liquid and a dark red liquid" instantly suspicious

>Not everyone will have the ability to wear them at all times

That's the point.

>Make exceptions for the chef

To be clear, I'm spriting alternates for the chef and bartender that they can get as an upgrade - kind of like if they wanted science goggles for their job but instead of just asking for science goggles, they could get some nice role specific ones from science, not something they should expect to have roundstart. Chefs have a reagent scanner in their PDA and bartenders get a smart container (the shaker).

**Comment by DrSnips** [comment]
I didnt mean you see each chem I ment it as there are only one for each for each state of matter so three at most.

**Comment by Dacendeth** [comment]
The mutation that lets you drink more should let you identify booze chemicals at a glance.

**Comment by jknpj** [comment]
It's been too long already so i'll just slap derelict on this. Ping us in the coder discord when you want it reopened.

**Pull request closed without merging**